### PR TITLE
[COLLABORATION] Added file system level collaboration architecture

### PR DESCRIPTION
**Pull request by hkirat**

### CODE REVIEW

**Comment by hkirat** [comment]
In reference to [#2342](https://github.com/mozilla/thimble.mozilla.org/issues/2342)

File: src/editor/Collaboration.js [hunk 1]
```diff
@@ -79,42 +79,63 @@ define(function (require, exports, module) {
...
+    function _handleCodemirrorChange (params) {
         if(_changing) {
             return;
         }
+        var delta = params.delta;
+        var relPath = params.path;
+        var fullPath = StartupState.project("root") + relPath;
+        var cm = _getOpenCodemirrorInstance(fullPath);
+        if(!cm) {
+            var file = FileSystemEntry.getFileForPath(fullPath);
```
Comment (gideonthomas) [inline src/editor/Collaboration.js]:
```js
if(!codemirror) {
    return handleFileEvent(codemirror);
}

...
```

File: src/editor/Collaboration.js [hunk 2]
```diff
@@ -79,42 +79,63 @@ define(function (require, exports, module) {
...
+        var from = cm.posFromIndex(start);
         var to = from;
-        _codemirror.replaceRange(param, from, to);
+        cm.replaceRange(param, from, to);
+        console.log("writting to file which is open in editor for path" + fullPath);
         _changing = false;
     };
 
-    function triggerCodemirrorChange(changeList) {
+    function _getOpenCodemirrorInstance(fullPath) {
```
Comment (gideonthomas) [inline src/editor/Collaboration.js]: use the codemirror instance from editor.js instead.

File: src/editor/Collaboration.js [hunk 3]
```diff
@@ -79,42 +79,63 @@ define(function (require, exports, module) {
...
+            return doc._masterEditor._codeMirror;
+        }
+        return null;
+    }
+
+    function triggerCodemirrorChange(changeList, fullPath) {
         if(_changing) {
             return;
         }
+        var relPath = fullPath.replace(StartupState.project("root"), "");
```
Comment (gideonthomas) [inline src/editor/Collaboration.js]: use Filer.path

File: src/editor/Collaboration.js [hunk 4]
```diff
@@ -79,42 +79,63 @@ define(function (require, exports, module) {
...
         for(var i = 0; i<changeList.length; i++) {
```
Comment (gideonthomas) [inline src/editor/Collaboration.js]: use `changeList.forEach`

**Comment by hkirat** [comment]
Up for review @gideonthomas

File: src/editor/Collaboration.js [hunk 5]
```diff
@@ -91,50 +87,75 @@ define(function (require, exports, module) {
...
+    function triggerCodemirrorChange(changeList, fullPath) {
         if(_changing) {
             return;
         }
-        for(var i = 0; i<changeList.length; i++) {
-            _webrtc.sendToAll("codemirror-change", changeList[i]);
-        }
+        var relPath = Path.relative(StartupState.project("root"), fullPath);
+            changeList.forEach(function(change) {
+            _webrtc.sendToAll("codemirror-change", {
```
Comment (gideonthomas) [inline src/editor/Collaboration.js]: indent

**Pull request merged**